README.md
# GNU MCU Eclipse QEMU - the build scripts

These are the scripts and additional files required to build the
[GNU MCU Eclipse QEMU](https://github.com/gnu-mcu-eclipse/qemu).

There are two types of builds:

- local/native builds, which use the tools available on the 
  host machine; generally the binaries do not run on a different system 
  distribution/version; intended mostly for development purposes.
- distribution builds, which create the archives distributed as 
  binaries; expected to run on most modern systems.

The build scripts use the 
[xPack Build Box (XBB)](https://github.com/xpack/xpack-build-box), 
a set of elaborate build environments based on GCC 7.2 (Docker containers
for GNU/Linux and Windows or a custom Homebrew for MacOS).

## Repository URLs

- the [GNU MCU Eclipse QEMU](https://github.com/gnu-mcu-eclipse/qemu) 
Git remote URL to clone from is https://github.com/gnu-mcu-eclipse/qemu.git
- the [QEMU](https://www.qemu.org) Git remote URL is
git://git.qemu.org/qemu.git

Add a remote named `qemu`, and pull the QEMU master → master.

## Download the build scripts repo

The build script is available from GitHub and can be 
[viewed online](https://github.com/gnu-mcu-eclipse/qemu-build/blob/master/scripts/build-native.sh).

To download it, clone the 
[gnu-mcu-eclipse/qemu-build](https://github.com/gnu-mcu-eclipse/qemu-build) 
Git repo, including submodules. 

```console
$ curl -L https://github.com/gnu-mcu-eclipse/qemu-build/raw/master/scripts/git-clone.sh | bash
```

which issues the following two commands:

```console
$ rm -rf ~/Downloads/qemu-build.git
$ git clone --recurse-submodules https://github.com/gnu-mcu-eclipse/qemu-build.git \
  ~/Downloads/qemu-build.git
```

To use the `develop` branch of the build scripts, use:

```console
$ rm -rf ~/Downloads/qemu-build.git
$ git clone --recurse-submodules -b develop https://github.com/gnu-mcu-eclipse/qemu-build.git \
  ~/Downloads/qemu-build.git
```

## The `Work` folder

The script creates a temporary build `Work/qemu-${version}` folder in 
the user home. Although not recommended, if for any reasons you need to 
change the location of the `Work` folder, 
you can redefine `WORK_FOLDER_PATH` variable before invoking the script.

## How to run a local/native build

### Prerequisites

For the moment the native build scripts were tested only on Ubuntu 18 LTS and macOS.
Details on how to prepare the development environment are in the 
[macOS](https://github.com/xpack/xpack-build-box/tree/master/macos)
and [Ubuntu](https://github.com/xpack/xpack-build-box/tree/master/ubuntu)
page of the [XBB project](https://github.com/xpack/xpack-build-box).

The Windows binaries are cross compiled with mingw-w64; this works on 
Ubuntu 18 LTS, including when running it under WSL (Windows System for Linux).

### Build

To build a macOS or Ubuntu binary based on the latest sources, run the
script without any options:

```console
$ bash ~/Downloads/qemu-build.git/scripts/build-native.sh
```

To create the Windows binaries, use:

```console
$ bash ~/Downloads/qemu-build.git/scripts/build-native.sh --win
```

The result is in `Work/qemu-dev/${platform}-${arch}/install/qemu`, 
with the executable in the `bin` folder.

For development builds, use:

```console
$ bash ~/Downloads/qemu-build.git/scripts/build-native.sh --develop --debug
```

or, for Windows:

```console
$ bash ~/Downloads/qemu-build.git/scripts/build-native.sh --develop --debug --win
```

### Clean

To clean the folders in preparation for a new build:

```console
$ bash ~/Downloads/qemu-build.git/scripts/build-native.sh clean
```

```console
$ bash ~/Downloads/qemu-build.git/scripts/build-native.sh cleanlibs
```

```console
$ bash ~/Downloads/qemu-build.git/scripts/build-native.sh cleanall
```

Similarly for Windows:

```console
$ bash ~/Downloads/qemu-build.git/scripts/build-native.sh --win clean
```

```console
$ bash ~/Downloads/qemu-build.git/scripts/build-native.sh --win cleanlibs
```

```console
$ bash ~/Downloads/qemu-build.git/scripts/build-native.sh --win cleanall
```

### DEVELOP.md

More details on the development environment for QEMU are in the separate
[DEVELOP](https://github.com/gnu-mcu-eclipse/qemu/blob/gnuarmeclipse-dev/DEVELOP.md)
page.

## How to build distributions

### Prerequisites

The prerequisites are common to all binary builds. Please follow the 
instructions in the separate 
[Prerequisites for building binaries](https://gnu-mcu-eclipse.github.io/developer/build-binaries-prerequisites-xbb/) 
page and return when ready.

### Preload the Docker images

Docker does not require to explicitly download new images, but does this 
automatically at first use.

However, since the images used for this build are relatively large, it 
is recommended to load them explicitly before starting the build:

```console
$ bash ~/Downloads/qemu-build.git/scripts/build.sh preload-images
```

The result should look similar to:

```console
$ docker images
REPOSITORY          TAG                 IMAGE ID            CREATED             SIZE
ilegeul/centos32    6-xbb-v1            f695dd6cb46e        2 weeks ago         2.92GB
ilegeul/centos      6-xbb-v1            294dd5ee82f3        2 weeks ago         3.09GB
hello-world         latest              f2a91732366c        2 months ago        1.85kB
```

### Update git repos

To keep the development repository in sync with the original QEMU 
repository:

- checkout `master`
- pull from `qemu/master`
- checkout `gnu-mcu-eclipse-dev`
- merge `master`
- add a tag like `v2.8.0-3-20180512` after each public release (mind the 
inner version `-3-`)

### Prepare release

To prepare a new release, first determine the QEMU version 
(like `2.8.0-3`) and update the `scripts/VERSION` file. The format is 
`2.8.0-3`. The fourth digit is the GNU MCU Eclipse release number 
of this version.

Add a new set of definitions in the `scripts/container-build.sh`, with 
the versions of various components.

### Update CHANGELOG.txt

Check `qemu-build.git/CHANGELOG.txt` and add the new release.

### Build

Although it is perfectly possible to build all binaries in a single step 
on a macOS system, due to Docker specifics, it is faster to build the 
GNU/Linux and Windows binaries on a GNU/Linux system and the macOS binary 
separately.

#### Build the GNU/Linux and Windows binaries

The current platform for GNU/Linux and Windows production builds is an 
Ubuntu 17.10 VirtualBox image running on a macMini with 16 GB of RAM 
and a fast SSD.

Before starting a multi-platform build, check if Docker is started:

```console
$ docker info
```

To build both the 32/64-bit Windows and GNU/Linux versions, use `--all`; 
to build selectively, use `--linux64 --win64` or `--linux32 --win32` 
(GNU/Linux can be built alone; Windows also requires the GNU/Linux build).

```console
$ rm -rf "${HOME}/Work"/qemu-*
$ bash ~/Downloads/qemu-build.git/scripts/build.sh --all
```

To build one of the previous versions:

```console
$ RELEASE_VERSION=2.8.0-3 bash ~/Downloads/qemu-build.git/scripts/build.sh --all
```

Several tens of minutes later, the output of the build script is a set of 
4 files and their SHA signatures, created in the `deploy` folder:

```console
$ ls -l deploy
total 30536
-rw-r--r-- 1 ilg ilg 7820067 May 23 10:45 gnu-mcu-eclipse-qemu-2.8.0-3-20180523-0703-centos32.tgz
-rw-r--r-- 1 ilg ilg     122 May 23 10:45 gnu-mcu-eclipse-qemu-2.8.0-3-20180523-0703-centos32.tgz.sha
-rw-r--r-- 1 ilg ilg 7505548 May 23 10:17 gnu-mcu-eclipse-qemu-2.8.0-3-20180523-0703-centos64.tgz
-rw-r--r-- 1 ilg ilg     122 May 23 10:17 gnu-mcu-eclipse-qemu-2.8.0-3-20180523-0703-centos64.tgz.sha
-rw-r--r-- 1 ilg ilg 7818371 May 23 11:00 gnu-mcu-eclipse-qemu-2.8.0-3-20180523-0703-win32.zip
-rw-r--r-- 1 ilg ilg     119 May 23 11:00 gnu-mcu-eclipse-qemu-2.8.0-3-20180523-0703-win32.zip.sha
-rw-r--r-- 1 ilg ilg 8026132 May 23 10:33 gnu-mcu-eclipse-qemu-2.8.0-3-20180523-0703-win64.zip
-rw-r--r-- 1 ilg ilg     119 May 23 10:33 gnu-mcu-eclipse-qemu-2.8.0-3-20180523-0703-win64.zip.sha
```

To copy the files from the build machine to the current development 
machine, open the `deploy` folder in a terminal and use `scp`:

```console
$ scp * ilg@ilg-mbp.local:Downloads/gme-binaries/qemu
```

#### Build the macOS binary

The current platform for macOS production builds is a macOS 10.10.5 
VirtualBox image running on the same macMini with 16 GB of RAM and a 
fast SSD.

To build the latest macOS version, with the same timestamp as the 
previous build:

```console
$ rm -rf "${HOME}/Work"/qemu-*
$ caffeinate bash ~/Downloads/qemu-build.git/scripts/build.sh --osx --date YYYYMMDD-HHMM
```

To build one of the previous macOS versions:

```console
$ RELEASE_VERSION=2.8.0-3 caffeinate bash ~/Downloads/qemu-build.git/scripts/build.sh --osx --date YYYYMMDD-HHMM
```

For consistency reasons, the date should be the same as the GNU/Linux 
and Windows builds.

Several minutes later, the output of the build script is a compressed 
archive and its SHA signature, created in the `deploy` folder:

```console
$ ls -l deploy
total 14152
-rw-r--r--  1 ilg  staff  7240894 May 23 16:07 gnu-mcu-eclipse-qemu-2.8.0-3-20180523-0703-macos.tgz
-rw-r--r--  1 ilg  staff      119 May 23 16:07 gnu-mcu-eclipse-qemu-2.8.0-3-20180523-0703-macos.tgz.sha
```

To copy the files from the build machine to the current development 
machine, open the `deploy` folder in a terminal and use `scp`:

```console
$ scp * ilg@ilg-mbp.local:Downloads/gme-binaries/qemu
```

### Subsequent runs

#### Separate platform specific builds

Instead of `--all`, you can use any combination of:

```
--win32 --win64 --linux32 --linux64
```

#### clean

To remove most build temporary files, use:

```console
$ bash ~/Downloads/qemu-build.git/scripts/build.sh --all clean
```

To also remove the library build temporary files, use:

```console
$ bash ~/Downloads/qemu-build.git/scripts/build.sh --all cleanlibs
```

To remove all temporary files, use:

```console
$ bash ~/Downloads/qemu-build.git/scripts/build.sh --all cleanall
```

Instead of `--all`, any combination of `--win32 --win64 --linux32 --linux64`
will remove the more specific folders.

For production builds it is recommended to completely remove the build folder.

#### --develop

For performance reasons, the actual build folders are internal to each 
Docker run, and are not persistent. This gives the best speed, but has 
the disadvantage that interrupted builds cannot be resumed.

For development builds, it is possible to define the build folders in 
the host file system, and resume an interrupted build.

#### --debug

For development builds, it is also possible to create everything with 
`-g -O0` and be able to run debug sessions.

#### Interrupted builds

The Docker scripts run with root privileges. This is generally not a 
problem, since at the end of the script the output files are reassigned 
to the actual user.

However, for an interrupted build, this step is skipped, and files in 
the install folder will remain owned by root. Thus, before removing 
the build folder, it might be necessary to run a recursive `chown`.

## Install

The procedure to install GNU MCU Eclipse QEMU is platform specific, 
but relatively straight forward (a .zip archive on Windows, a compressed 
tar archive on macOS and GNU/Linux).

A portable method is to use [`xpm`](https://www.npmjs.com/package/xpm):

```console
$ xpm install --global @gnu-mcu-eclipse/qemu
```

More details are available on the 
[How to install the QEMU binaries?](https://gnu-mcu-eclipse.github.io/qemu/install/)
page.

After install, the package should create a structure like this (only the 
first two depth levels are shown):

```console
$ tree -L 2 /Users/ilg/Library/xPacks/\@gnu-mcu-eclipse/qemu/xxxx/.content/
/Users/ilg/Library/xPacks/\@gnu-mcu-eclipse/qemu/xxxx/.content/
...
```

No other files are installed in any system folders or other locations.

## Uninstall

The binaries are distributed as portable archives; thus they do not need 
to run a setup and do not require an uninstall; simply removing the
folder is enough.

## Actual configuration

The result of the `configure` step on CentOS 6, with most of the 
options disabled, is:

```
Source path       /Host/Work/qemu-2.8.0-4/qemu.git
C compiler        gcc
Host C compiler   cc
C++ compiler      g++
Objective-C compiler gcc
ARFLAGS           rv
CFLAGS            -O2 -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=2 -g -ffunction-sections -fdata-sections -m64 -pipe -O2 -Wno-format-truncation -Wno-incompatible-pointer-types -Wno-unused-function -Wno-unused-but-set-variable -Wno-unused-result
QEMU_CFLAGS       -I/Host/Work/qemu-2.8.0-4/install/centos64/include/pixman-1 -I$(SRC_PATH)/dtc/libfdt -pthread -I/Host/Work/qemu-2.8.0-4/install/centos64/include/glib-2.0 -I/Host/Work/qemu-2.8.0-4/install/centos64/lib/glib-2.0/include -fPIE -DPIE -m64 -mcx16 -D_GNU_SOURCE -D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE -Wstrict-prototypes -Wredundant-decls -Wall -Wundef -Wwrite-strings -Wmissing-prototypes -fno-strict-aliasing -fno-common -fwrapv  -ffunction-sections -fdata-sections -m64 -pipe -O2 -Wno-format-truncation -Wno-incompatible-pointer-types -Wno-unused-function -Wno-unused-but-set-variable -Wno-unused-result -I/Host/Work/qemu-2.8.0-4/install/centos64/include -Wendif-labels -Wno-shift-negative-value -Wmissing-include-dirs -Wempty-body -Wnested-externs -Wformat-security -Wformat-y2k -Winit-self -Wignored-qualifiers -Wold-style-declaration -Wold-style-definition -Wtype-limits -fstack-protector-strong
LDFLAGS           -Wl,--warn-common -Wl,-z,relro -Wl,-z,now -pie -m64 -g -L/Host/Work/qemu-2.8.0-4/install/centos64/lib -L/Host/Work/qemu-2.8.0-4/install/centos64/lib
make              make
install           install
python            python -B
module support    no
host CPU          x86_64
host big endian   no
target list       gnuarmeclipse-softmmu
tcg debug enabled yes
gprof enabled     no
sparse enabled    no
strip binaries    no
profiler          no
static build      no
pixman            system
SDL support       yes (2.0.5)
GTK support       no 
GTK GL support    no
VTE support       no 
TLS priority      NORMAL
GNUTLS support    no
GNUTLS rnd        no
libgcrypt         no
libgcrypt kdf     no
nettle            no 
nettle kdf        no
libtasn1          no
curses support    no
virgl support     no
curl support      no
mingw32 support   no
Audio drivers     
Block whitelist (rw) 
Block whitelist (ro) 
VirtFS support    no
VNC support       no
xen support       no
brlapi support    no
bluez  support    no
Documentation     yes
PIE               yes
vde support       no
netmap support    no
Linux AIO support no
ATTR/XATTR support yes
Install blobs     no
KVM support       no
COLO support      yes
RDMA support      no
TCG interpreter   no
fdt support       yes
preadv support    yes
fdatasync         yes
madvise           yes
posix_madvise     yes
libcap-ng support no
vhost-net support yes
vhost-scsi support yes
vhost-vsock support yes
Trace backends    log
spice support     no 
rbd support       no
xfsctl support    no
smartcard support no
libusb            no
usb net redir     no
OpenGL support    no
OpenGL dmabufs    no
libiscsi support  no
libnfs support    no
build guest agent no
QGA VSS support   no
QGA w32 disk info no
QGA MSI support   no
seccomp support   no
coroutine backend ucontext
coroutine pool    yes
debug stack usage no
GlusterFS support no
Archipelago support no
gcov              gcov
gcov enabled      no
TPM support       no
libssh2 support   no
TPM passthrough   no
QOM debugging     yes
lzo support       no
snappy support    no
bzip2 support     no
NUMA host support no
tcmalloc support  no
jemalloc support  no
avx2 optimization yes
replication support yes
```

## Test

A simple test is performed by the script at the end, by launching the 
executable to check if all shared/dynamic libraries are correctly used.

For a true test you need to first install the package and then run the 
program from the final location. For example on macOS the output should 
look like:

```console
$ /Users/ilg/Library/xPacks/\@gnu-mcu-eclipse/qemu/xxxx/.content/bin/qemu --version
GNU MCU Eclipse 64-bit QEMU ...
```

## More build details

The build process is split into several scripts. The build starts on 
the host, with `build.sh`, which runs `container-build.sh` several 
times, once for each target, in one of the two docker containers. 
Both scripts include several other helper scripts. The entire process 
is quite complex, and an attempt to explain its functionality in a few 
words would not be realistic. Thus, the authoritative source of details 
remains the source code.


CHANGELOG.txt

The following changes (in reverse order) were applied to the QEMU master branch:

[2019-04-24]

- v2.8.0-5 20190424 released

[2019-04-10]

- fix STM32F4 SR reset value

[2019-02-26]

- [#63] cortexm/mcu.c: use '-m size=kb'; use the value from the machine object, which can be overriden by command line options.
- [#43] arm-semi.c: SYS_HEAPINFO returns all 0
- [#39] vl.c: fix crash for unsupported board
- add .vscode with build tasks and debug launchers
- Makefile: add install-gme

[2019-02-11]

- v2.8.0-4-20190211 released
- [#62] fix the GPIO persistence issue
- the greeting shows 32-bit or 64-bit (singular).

[2018-05-23]

- v2.8.0-3-20180523 released
- use new build scripts based on XBB

[2017-06-15]

- rebranded as GNU MCU Eclipse QEMU

[2016-12-27]

- [#24] vl.c: fix semihosting parsing of other opts

[2016-12-26]

- STM32-P407 with functional buttons
- OLIMEXINO-STM32 with functional buttons
- STM32-P107 with functional buttons
- Netduino plus 2, Netduino Go & Maple with functional buttons
- merge master 2.8.0

[2016-12-25]

- NUCLEO-F411RE with functional buttons
- STM32-P103 with functional buttons

[2016-12-24]

- stm32/gpio: fix exti check
- NUCLEO-F103RB with functional buttons 
- stm32: afio & syscfg with enable bit

[2016-12-22]

- stm32f429i-discovery with functional buttons 

[2016-12-18]

- stm32: rearrange the `realize()` logic
- add `peripheral_create_memory_region()`
- add `peripheral_prepare_registers()`

[2016-12-17]

- svd: generate separate files for peripherals
- qemu.nsi: add devices folder to setup

[2016-12-15]

- stm32/afio added; F1 gpio uses it
- stm32 & cortexm reset all children devices
- add scripts to process SVD files
- STM32-H103 with functional buttons 

[2016-12-13]

- cortexm/stm32: use union to group families

[2016-12-12]

- stm32/gpio redefined with SVD values only

[2016-12-03]

- rename images -> graphics

[2016-12-01]

- add STM32F0-Discovery, with buttons and leds

[2016-11-28]

- remove `armv7m_nvic.o` from the build; refer to `cortexm_nvic_*` in `helper.c`

[2016-11-24]

- rename `LOG_FUNC`, `LOG_MR` (`-d func,mr`)

[2016-11-24]

- stm32: add EXTI, SYSCFG
- stm32/capabilities: add IRQn defs
- add support for user buttons

[2016-11-16]

- button-reset: actions functional

[2016-11-16]

- cortexm: add graphic buttons, reset & user; not linked to actions (yet)

[2016-11-09]

- qemu-thread-posix: `PTHREAD_MUTEX_ERRORCHECK`

[2016-11-02]

- add support for SDL2

[2016-10-29]

- version 2.7.0-20161029 released on GitHub
- gnuarmeclipse-dev merged to gnuarmeclipse

[2016-10-26]

- avoid `clock_gettime()` on Apple
- fix `fix cm_cpu_generic_create()` bug

[2016-10-24]

- original version 2.7.0 merged to gnuarmeclipse-dev
- add separate cortexm-bitband implementation

[2016-10-20]

- nsi file: add InstallDir

[2016-07-28]

- version 2.6.0-20160728 released on GitHub
- fix half word writes (register_post_write_callback_t)
- implement (minimally) the DHCSR register, for C_DEBUGEN
- consider BASEPRI for disabling interrupts
- add more registers to GDB server (MSP, PSP, PRIMASK, BASEPRI, FAULTMASK, CONTROL)

[2016-07-19]

- peripheral-register: fix reset
- add reset for all uarts

[2015-10-29]

- version 2.4.50-20151029, released on GitHub
- fix rendering on OS X 10.11 (SDL)

[2015-08-16]

- version 2.3.50-20150816*-dev released
- add build for Win64

[2015-08-04]

- version 2.3.50-20150804*-dev released
- SDL event loop added (to fix responsivness & Windows bug)
- stm32f411re added
- nucleo-f411re board added

[2015-08-01]

- version 2.3.50-20150801*-dev released

[2015-07-23]

- qemu-options.hx: -board, -mcu added
- vl.c: -board & -mcu parsed
- null-machine.c disabled
- cortexm-board.c added ('generic')

[2015-07-16]

- build: add SDL
- gpio-led: add graphical blink support

[2015-06-25]

- cortex-mcu: use *_do_unassigned_access_callback; currently just display
a message, no exceptions thrown.

[2015-06-22]

- cortexm-nvic added

[2015-06-21]

- add '--image filename.elf'

[2015-06-17]

- stm32: add F2, F3, L1 families
- /machine/cortexm container added; nvic & itm inside
- /machine/stm32 container; rcc, flash, gpio[%c] in

[2015-06-10]

- cortexm-mcu: properties *-size-kb renamed
- stm32-mcu: add hsi-freq-hz & lsi-freq-hz props
- stm32-rcc: update clock functional

[2015-06-09]

- cpu.c: log MSP & PC

[2015-06-08]

- '-d trace_mr' (LOG_TRACE) added to log; intended for development use
- loader.c: add verbosity for loaded segments
- loader: make rom_reset public
- cortexm: do a rom_reset() before cpu_reset()

[2015-06-02]

- '-d trace' (LOG_TRACE) added to log; intended for development use

[2015-05-12]

- the build scripts running in virtual machines were deprecated, and a single script,
using Docker, was added to the main gnuarmeclipse-se.git/scripts.

- the greeting shows 32-bits or 64-bits (plural for bits).


[2015-01-20]

- build script for OS X added.

[2014-11-30]

- custom definitions for the Windows setup creator.

[2014-11-25]

- sam & xmc added boards added

[2014-11-24]

- set default system_clock_scale = 80

[2014-11-20]

- cortex-m intial implementation
- most stm32 boards added
- Kinetis & LPC boards added
- tiva board added

[2014-11-04 to 08]

- semihosting fixed
- verbosity added, including the connection message, required by plug-in
- branding added


Liviu Ionescu


scripts/build.sh
#!/usr/bin/env bash
# -----------------------------------------------------------------------------
# This file is part of the GNU MCU Eclipse distribution.
#   (https://gnu-mcu-eclipse.github.io)
# Copyright (c) 2019 Liviu Ionescu.
#
# Permission to use, copy, modify, and/or distribute this software 
# for any purpose is hereby granted, under the terms of the MIT license.
# -----------------------------------------------------------------------------

# -----------------------------------------------------------------------------
# Safety settings (see https://gist.github.com/ilg-ul/383869cbb01f61a51c4d).

if [[ ! -z ${DEBUG} ]]
then
  set ${DEBUG} # Activate the expand mode if DEBUG is anything but empty.
else
  DEBUG=""
fi

set -o errexit # Exit if command failed.
set -o pipefail # Exit if pipe failed.
set -o nounset # Exit if variable not set.

# Remove the initial space and instead use '\n'.
IFS=$'\n\t'

# -----------------------------------------------------------------------------
# Identify the script location, to reach, for example, the helper scripts.

build_script_path="$0"
if [[ "${build_script_path}" != /* ]]
then
  # Make relative path absolute.
  build_script_path="$(pwd)/$0"
fi

script_folder_path="$(dirname "${build_script_path}")"
script_folder_name="$(basename "${script_folder_path}")"

# =============================================================================

# Script to build the GNU MCU Eclipse ARM QEMU distribution packages.
#
# Developed on macOS 10.13 High Sierra, but intended to run on
# macOS 10.10 Yosemite and CentOS 6 XBB. 

# -----------------------------------------------------------------------------

echo
echo "GNU MCU Eclipse QEMU distribution build script."

host_functions_script_path="${script_folder_path}/helper/host-functions-source.sh"
source "${host_functions_script_path}"

# -----------------------------------------------------------------------------

# Array where the remaining args will be stored.
declare -a rest

help_message="    bash $0 [--win32] [--win64] [--linux32] [--linux64] [--osx] [--all] [clean|cleanlibs|cleanall|preload-images] [--env-file file] [--date YYYYmmdd-HHMM] [--disable-strip] [--without-pdf] [--with-html] [--develop] [--debug] [--jobs N] [--help]"
host_options "${help_message}" $@

echo
echo "Host helper functions source script: \"${host_functions_script_path}\"."

host_detect

# docker_linux64_image="ilegeul/centos:6-xbb-v2.1"
# docker_linux32_image="ilegeul/centos32:6-xbb-v2.1"

host_common

CONTAINER_RUN_AS_ROOT="y"

# -----------------------------------------------------------------------------

if [ -n "${DO_BUILD_WIN32}${DO_BUILD_WIN64}${DO_BUILD_LINUX32}${DO_BUILD_LINUX64}" ]
then
  host_prepare_docker
fi

# ----- Build the native distribution. ----------------------------------------

if [ -z "${DO_BUILD_ANY}" ]
then

  host_build_target "Creating the native distribution..." \
    --script "${HOST_WORK_FOLDER_PATH}/${CONTAINER_BUILD_SCRIPT_REL_PATH}" \
    --env-file "${ENV_FILE}" \
    -- \
    ${rest[@]-}

else

  # ----- Build the OS X distribution. ----------------------------------------

  if [ "${DO_BUILD_OSX}" == "y" ]
  then
    if [ "${HOST_UNAME}" == "Darwin" ]
    then
      host_build_target "Creating the OS X distribution..." \
        --script "${HOST_WORK_FOLDER_PATH}/${CONTAINER_BUILD_SCRIPT_REL_PATH}" \
        --env-file "${ENV_FILE}" \
        --target-platform "darwin" \
        -- \
        ${rest[@]-}
    else
      echo "Building the macOS image is not possible on this platform."
      exit 1
    fi
  fi

  # ----- Build the GNU/Linux 64-bit distribution. ---------------------------

  if [ "${DO_BUILD_LINUX64}" == "y" ]
  then
    host_build_target "Creating the GNU/Linux 64-bit distribution..." \
      --script "${CONTAINER_WORK_FOLDER_PATH}/${CONTAINER_BUILD_SCRIPT_REL_PATH}" \
      --env-file "${ENV_FILE}" \
      --target-platform "linux" \
      --target-arch "x64" \
      --target-bits 64 \
      --docker-image "${docker_linux64_image}" \
      -- \
      ${rest[@]-}
  fi

  # ----- Build the Windows 64-bit distribution. -----------------------------

  if [ "${DO_BUILD_WIN64}" == "y" ]
  then
    host_build_target "Creating the Windows 64-bit distribution..." \
      --script "${CONTAINER_WORK_FOLDER_PATH}/${CONTAINER_BUILD_SCRIPT_REL_PATH}" \
      --env-file "${ENV_FILE}" \
      --target-platform "win32" \
      --target-arch "x64" \
      --target-bits 64 \
      --docker-image "${docker_linux64_image}" \
      -- \
      ${rest[@]-}
  fi

  # ----- Build the GNU/Linux 32-bit distribution. ---------------------------

  if [ "${DO_BUILD_LINUX32}" == "y" ]
  then
    host_build_target "Creating the GNU/Linux 32-bit distribution..." \
      --script "${CONTAINER_WORK_FOLDER_PATH}/${CONTAINER_BUILD_SCRIPT_REL_PATH}" \
      --env-file "${ENV_FILE}" \
      --target-platform "linux" \
      --target-arch "x32" \
      --target-bits 32 \
      --docker-image "${docker_linux32_image}" \
      -- \
      ${rest[@]-}
  fi

  # ----- Build the Windows 32-bit distribution. -----------------------------

  # Since the actual container is a 32-bit, use the debian32 binaries.
  if [ "${DO_BUILD_WIN32}" == "y" ]
  then
    host_build_target "Creating the Windows 32-bit distribution..." \
      --script "${CONTAINER_WORK_FOLDER_PATH}/${CONTAINER_BUILD_SCRIPT_REL_PATH}" \
      --env-file "${ENV_FILE}" \
      --target-platform "win32" \
      --target-arch "x32" \
      --target-bits 32 \
      --docker-image "${docker_linux32_image}" \
      -- \
      ${rest[@]-}
  fi

fi

host_show_sha

# -----------------------------------------------------------------------------

host_stop_timer

host_notify_completed

echo
echo "Use --date ${DISTRIBUTION_FILE_DATE} if needed for a related build."

# Completed successfully.
exit 0

# -----------------------------------------------------------------------------
